### ISSUE TITLE
[Bug]: Wrong tick labels and colorbar of discrete normalizer

### ISSUE THREAD

**Issue by zxdawn** [matplotlib/matplotlib#21841]
### Bug summary

The discrete normalizer produces duplicated labels for the same tick and the colorbar is wrong.

### Code for reproduction

```python
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
import matplotlib.ticker as mticker

fig, ax = plt.subplots()

edges = np.linspace(0, 3, 13)
norm = mcolors.BoundaryNorm(edges, 1000)
Z = np.random.rand(5, 5) * 3

m = ax.pcolormesh(Z, cmap='viridis', norm=norm)
cb = fig.colorbar(m, ticks=mticker.MultipleLocator(0.25))

fig.show()
```

### Actual outcome

![image](https://user-images.githubusercontent.com/30388627/144408972-91e06931-6ae0-4a0c-8d48-bf4274bed2aa.png)

### Expected outcome

Correct ticks and colorbar.

### Operating system

Ubuntu

### Matplotlib Version

3.5.0

### Matplotlib Backend

module://matplotlib_inline.backend_inline

### Python version

Python 3.9.7

### Installation

conda

**Comment by jklymak** [comment]
Did this work differently before?

**Comment by zxdawn** [comment]
For `3.2.2`, the ticks are correct:
![image](https://user-images.githubusercontent.com/30388627/144411161-69566e23-8bbc-448c-9159-242f57782f64.png)

**Comment by jklymak** [comment]
OK confirm this bug on 3.5.0.  It has however been fixed by #21676 and should be fixed in 3.5.1.  Sorry for the inconvenience!
